File: src/main/java/com/cxxcxx/zinecraft/api/datagen/CatalogLanguageProvider.java
package com.cxxcxx.zinecraft.api.datagen;

import com.cxxcxx.zinecraft.api.registry.catalog.TranslationCatalog;
import net.minecraft.data.PackOutput;
import net.neoforged.neoforge.common.data.LanguageProvider;

/**
 * Emits one locale from the shared translation catalog.
 */
public final class CatalogLanguageProvider extends LanguageProvider {
  private final TranslationCatalog translations;
  private final String locale;

  public CatalogLanguageProvider(PackOutput output, TranslationCatalog translations, String locale) {
    super(output, "zinecraft", locale);
    this.translations = translations;
    this.locale = locale;
  }

  @Override
  protected void addTranslations() {
    translations.getEntries().forEach((key, text) ->
        add(key, locale.equals("zh_cn") ? text.getZhCn() : text.getEnUs())
    );
  }
}

File: src/main/kotlin/com/cxxcxx/zinecraft/api/datagen/CatalogLanguageProvider.kt
package com.cxxcxx.zinecraft.api.datagen

import com.cxxcxx.zinecraft.api.localization.TranslationCatalog
import net.fabricmc.fabric.api.datagen.v1.FabricDataOutput
import net.fabricmc.fabric.api.datagen.v1.provider.FabricLanguageProvider
import net.minecraft.core.HolderLookup
import java.util.concurrent.CompletableFuture

class CatalogLanguageProvider(
  output: FabricDataOutput,
  registries: CompletableFuture<HolderLookup.Provider?>?,
  private val translations: TranslationCatalog,
  private val locale: String
) : FabricLanguageProvider(output, locale, registries) {
  override fun generateTranslations(
    holderLookup: HolderLookup.Provider?,
    translationBuilder: TranslationBuilder?
  ) {
    translations.entries.forEach { (key, text) ->
      translationBuilder?.add(key, if (locale == "zh_cn") text.zhCn else text.enUs)
    }
  }
}
